### DISCUSSION TITLE
allow optimisticData to return undefined

### DISCUSSION THREAD

**Discussion by samsch** [vercel/swr#2794] [Ideas]
the function passed as `optimisticData` in mutate gets two parameters, where `currentData` can explicitly be undefined regardless of `Data` (most recent updates from this PR: https://github.com/vercel/swr/pull/2708/files#diff-947adbc8aa4f84c3191ad83c84d1344a1fd311224f73a9f92e7b9eaa73860508R324). if you are using the function version of `optimisticData` though, it's very possible you can't create the whole object expected, so can we explicitly allow the return value of `optimisticData` to be undefined?

Right now this gets a type error because `optimisticData` infers that the return value should be `DataType`.
```ts
mutate('/somekey', undefined, {
	optimisticData: (previous?: DataType) => {
		if (previous == null) {
			return undefined;
		}
		return {
			...previous,
			updatedProperty: newValue,
		};
	},
});
```

It can be worked around by changing the first line to explicitly have the data type or undefined
```ts
mutate<DataType | undefined>('/somekey', undefined, {
```
The addition of an explicit type or undefined lets this work. but the value of swr.data is already allowed to be undefined.

Can we have optimisticData's inferred type also explicitly allow returning `undefined`?

**Comment by dmitriiprusakov** [comment]
@samsch Same question!
